File: app/src/main/java/controllers/MagicCardRetrofitSingleton.java
package controllers;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class MagicCardRetrofitSingleton {
    static final String BASE_URL = "http://10.0.2.2:8888/";

    private static InterfaceMagicCardAPI magicCardAPIInstance;

    public static InterfaceMagicCardAPI getInstance (){
        if ( magicCardAPIInstance== null)
            synchronized (InterfaceMagicCardAPI.class){
                createApiBuilder();
            }
        return magicCardAPIInstance;
    }

    private static void createApiBuilder() {
        Gson gson = new GsonBuilder()
                .setLenient()
                .create();

        Retrofit retrofit = new Retrofit.Builder()
                .baseUrl(BASE_URL)
                .addConverterFactory(GsonConverterFactory.create(gson))
                .build();
        magicCardAPIInstance = retrofit.create(InterfaceMagicCardAPI.class);
    }
}

File: app/src/main/java/controllers/MagicCardRetrofitSingleton.kt
package controllers

import com.google.gson.GsonBuilder
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

object MagicCardRetrofitSingleton {
    internal val BASE_URL = "http://17ruecroixberthet.freeboxos.fr/"
  //internal val BASE_URL = "http://10.0.2.2:8888/"



    private var magicCardAPIInstance: InterfaceMagicCardAPI? = null

    val instance: InterfaceMagicCardAPI?
        get() {
            if (magicCardAPIInstance == null)
                createApiBuilder()

            return magicCardAPIInstance
        }


    private val interceptorLog = HttpLoggingInterceptor().let {
        it.level = HttpLoggingInterceptor.Level.BODY
        it
    }

    private fun createApiBuilder() {
        val gson = GsonBuilder()
                .setLenient()
                .create()


        //OkhttpClient for building http request url
         val okHttpClient = OkHttpClient.Builder()
                .addInterceptor(interceptorLog)
                .build()

        val retrofit = Retrofit.Builder()
                .client(okHttpClient)
                .baseUrl(BASE_URL)
                .addConverterFactory(GsonConverterFactory.create(gson))
                .build()
        magicCardAPIInstance = retrofit.create(InterfaceMagicCardAPI::class.java)
    }
}
